File: migration/src/storage/sqlite.ts
/**
 * SQLite Storage Adapter
 *
 * Stores news items in a SQLite database.
 * Uses better-sqlite3 for synchronous API.
 *
 * NOTE: Requires `better-sqlite3` package:
 *   npm install better-sqlite3
 *   npm install -D @types/better-sqlite3
 */

import * as fs from 'fs';
import * as path from 'path';
import type { NormalizedNewsItem, StorageAdapter } from '../types';
import { logger } from '../utils';

// Type for better-sqlite3 (optional dependency)
interface Database {
  prepare(sql: string): Statement;
  exec(sql: string): void;
  close(): void;
}

interface Statement {
  run(...params: unknown[]): void;
  get(...params: unknown[]): unknown;
  all(...params: unknown[]): unknown[];
}

export interface SqliteStorageOptions {
  /** Database file path */
  filePath: string;
  /** Table name (default: 'news_items') */
  tableName?: string;
}

const CREATE_TABLE_SQL = `
  CREATE TABLE IF NOT EXISTS news_items (
    id TEXT PRIMARY KEY,
    source TEXT NOT NULL,
    url TEXT NOT NULL,
    title TEXT NOT NULL,
    published_at TEXT NOT NULL,
    fetched_at TEXT NOT NULL,
    authors TEXT,
    summary TEXT,
    content_text TEXT,
    tickers TEXT,
    topics TEXT,
    language TEXT DEFAULT 'en',
    metadata TEXT,
    created_at TEXT DEFAULT CURRENT_TIMESTAMP
  )
`;

const CREATE_INDEX_SQL = `
  CREATE INDEX IF NOT EXISTS idx_published_at ON news_items(published_at);
  CREATE INDEX IF NOT EXISTS idx_source ON news_items(source);
`;

/**
 * SQLite Storage Adapter
 */
export class SqliteStorage implements StorageAdapter {
  private filePath: string;
  private tableName: string;
  private db: Database | null = null;

  constructor(options: SqliteStorageOptions) {
    this.filePath = options.filePath;
    this.tableName = options.tableName ?? 'news_items';

    // Ensure output directory exists
    const dir = path.dirname(this.filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  private async getDb(): Promise<Database> {
    if (this.db) return this.db;

    try {
      // Dynamic import for optional dependency
      const BetterSqlite3 = await import('better-sqlite3');
      const Database = BetterSqlite3.default || BetterSqlite3;
      this.db = new Database(this.filePath) as Database;

      // Create table if not exists
      this.db.exec(CREATE_TABLE_SQL);
      this.db.exec(CREATE_INDEX_SQL);

      return this.db;
    } catch (error) {
      throw new Error(
        `SQLite storage requires 'better-sqlite3' package. Install with: npm install better-sqlite3\n` +
        `Original error: ${(error as Error).message}`
      );
    }
  }

  async save(items: NormalizedNewsItem[]): Promise<void> {
    const db = await this.getDb();

    const insertSql = `
      INSERT OR REPLACE INTO news_items
      (id, source, url, title, published_at, fetched_at, authors, summary, content_text, tickers, topics, language, metadata)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `;

    const stmt = db.prepare(insertSql);

    for (const item of items) {
      stmt.run(
        item.id,
        item.source,
        item.url,
        item.title,
        item.published_at,
        item.fetched_at,
        JSON.stringify(item.authors),
        item.summary,
        item.content_text,
        JSON.stringify(item.tickers),
        JSON.stringify(item.topics),
        item.language,
        JSON.stringify(item.metadata)
      );
    }

    logger.info(`SQLite: Saved ${items.length} items to ${this.filePath}`);
  }

  async load(): Promise<NormalizedNewsItem[]> {
    const db = await this.getDb();

    const rows = db.prepare(`SELECT * FROM news_items ORDER BY published_at DESC`).all() as Array<{
      id: string;
      source: string;
      url: string;
      title: string;
      published_at: string;
      fetched_at: string;
      authors: string;
      summary: string;
      content_text: string;
      tickers: string;
      topics: string;
      language: string;
      metadata: string;
    }>;

    const items: NormalizedNewsItem[] = rows.map((row) => ({
      id: row.id,
      source: row.source,
      url: row.url,
      title: row.title,
      published_at: row.published_at,
      fetched_at: row.fetched_at,
      authors: JSON.parse(row.authors || '[]'),
      summary: row.summary || '',
      content_text: row.content_text || '',
      tickers: JSON.parse(row.tickers || '[]'),
      topics: JSON.parse(row.topics || '[]'),
      language: row.language || 'en',
      metadata: JSON.parse(row.metadata || '{}'),
    }));

    logger.info(`SQLite: Loaded ${items.length} items from ${this.filePath}`);
    return items;
  }

  async clear(): Promise<void> {
    const db = await this.getDb();
    db.exec(`DELETE FROM news_items`);
    logger.info(`SQLite: Cleared all items from ${this.filePath}`);
  }

  /**
   * Query items with SQL WHERE clause
   */
  async query(whereClause: string, params: unknown[] = []): Promise<NormalizedNewsItem[]> {
    const db = await this.getDb();
    const sql = `SELECT * FROM news_items WHERE ${whereClause} ORDER BY published_at DESC`;
    const rows = db.prepare(sql).all(...params) as Array<Record<string, unknown>>;

    return rows.map((row) => ({
      id: row.id as string,
      source: row.source as string,
      url: row.url as string,
      title: row.title as string,
      published_at: row.published_at as string,
      fetched_at: row.fetched_at as string,
      authors: JSON.parse((row.authors as string) || '[]'),
      summary: (row.summary as string) || '',
      content_text: (row.content_text as string) || '',
      tickers: JSON.parse((row.tickers as string) || '[]'),
      topics: JSON.parse((row.topics as string) || '[]'),
      language: (row.language as string) || 'en',
      metadata: JSON.parse((row.metadata as string) || '{}'),
    }));
  }

  /**
   * Get count of items
   */
  async count(): Promise<number> {
    const db = await this.getDb();
    const result = db.prepare(`SELECT COUNT(*) as count FROM news_items`).get() as { count: number };
    return result.count;
  }

  /**
   * Close database connection
   */
  close(): void {
    if (this.db) {
      this.db.close();
      this.db = null;
    }
  }

  /**
   * Get the database file path
   */
  getFilePath(): string {
    return this.filePath;
  }
}

/**
 * Create a SQLite storage adapter with default output path
 */
export function createSqliteStorage(outputDir = './output'): SqliteStorage {
  const filePath = path.join(outputDir, 'news.db');
  return new SqliteStorage({ filePath });
}

File: py_migration/news_scanner/storage/sqlite.py
"""
SQLite Storage Adapter

Stores news items in a SQLite database.
"""

import os
import json
import sqlite3
from pathlib import Path
from typing import Optional

from .base import BaseStorage
from ..models import NormalizedNewsItem, NewsMetadata
from ..utils import logger


CREATE_TABLE_SQL = """
CREATE TABLE IF NOT EXISTS news_items (
    id TEXT PRIMARY KEY,
    source TEXT NOT NULL,
    url TEXT NOT NULL,
    title TEXT NOT NULL,
    published_at TEXT NOT NULL,
    fetched_at TEXT NOT NULL,
    authors TEXT,
    summary TEXT,
    content_text TEXT,
    tickers TEXT,
    topics TEXT,
    language TEXT DEFAULT 'en',
    metadata TEXT,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
)
"""

CREATE_INDICES_SQL = """
CREATE INDEX IF NOT EXISTS idx_published_at ON news_items(published_at);
CREATE INDEX IF NOT EXISTS idx_source ON news_items(source);
CREATE INDEX IF NOT EXISTS idx_created_at ON news_items(created_at);
"""


class SqliteStorage(BaseStorage):
    """
    SQLite database storage.

    Example usage:
        storage = SqliteStorage("output/news.db")
        await storage.save(items)
        items = await storage.load()

        # Query with SQL:
        items = storage.query("source = ?", ["BBC World"])
        count = storage.count()
    """

    def __init__(self, db_path: str):
        """
        Initialize SQLite storage.

        Args:
            db_path: Path to SQLite database file.
        """
        self.db_path = db_path
        self._conn: Optional[sqlite3.Connection] = None

        # Ensure directory exists
        Path(db_path).parent.mkdir(parents=True, exist_ok=True)

        # Initialize database
        self._init_db()

    def _init_db(self) -> None:
        """Initialize database tables."""
        conn = self._get_conn()
        conn.executescript(CREATE_TABLE_SQL + CREATE_INDICES_SQL)
        conn.commit()

    def _get_conn(self) -> sqlite3.Connection:
        """Get database connection."""
        if self._conn is None:
            self._conn = sqlite3.connect(self.db_path)
            self._conn.row_factory = sqlite3.Row
        return self._conn

    async def save(self, items: list[NormalizedNewsItem]) -> None:
        """Save news items to database."""
        conn = self._get_conn()

        insert_sql = """
        INSERT OR REPLACE INTO news_items
        (id, source, url, title, published_at, fetched_at, authors, summary,
         content_text, tickers, topics, language, metadata)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """

        for item in items:
            conn.execute(insert_sql, (
                item.id,
                item.source,
                item.url,
                item.title,
                item.published_at,
                item.fetched_at,
                json.dumps(item.authors),
                item.summary,
                item.content_text,
                json.dumps(item.tickers),
                json.dumps(item.topics),
                item.language,
                json.dumps(item.metadata.to_dict() if item.metadata else {}),
            ))

        conn.commit()
        logger.info(f"SQLite: Saved {len(items)} items to {self.db_path}")

    async def load(self) -> list[NormalizedNewsItem]:
        """Load all news items from database."""
        conn = self._get_conn()
        cursor = conn.execute(
            "SELECT * FROM news_items ORDER BY published_at DESC"
        )

        items = []
        for row in cursor:
            items.append(self._row_to_item(row))

        logger.info(f"SQLite: Loaded {len(items)} items from {self.db_path}")
        return items

    async def clear(self) -> None:
        """Clear all items from database."""
        conn = self._get_conn()
        conn.execute("DELETE FROM news_items")
        conn.commit()
        logger.info(f"SQLite: Cleared {self.db_path}")

    def query(
        self,
        where_clause: str,
        params: Optional[list] = None,
        limit: Optional[int] = None
    ) -> list[NormalizedNewsItem]:
        """
        Query items with SQL WHERE clause.

        Args:
            where_clause: SQL WHERE clause (without "WHERE").
            params: Query parameters.
            limit: Max items to return.

        Returns:
            List of matching items.
        """
        conn = self._get_conn()

        sql = f"SELECT * FROM news_items WHERE {where_clause} ORDER BY published_at DESC"
        if limit:
            sql += f" LIMIT {limit}"

        cursor = conn.execute(sql, params or [])
        return [self._row_to_item(row) for row in cursor]

    def count(self) -> int:
        """Get count of items in database."""
        conn = self._get_conn()
        cursor = conn.execute("SELECT COUNT(*) FROM news_items")
        return cursor.fetchone()[0]

    def get_sources(self) -> list[str]:
        """Get list of unique sources."""
        conn = self._get_conn()
        cursor = conn.execute("SELECT DISTINCT source FROM news_items ORDER BY source")
        return [row[0] for row in cursor]

    def get_by_source(self, source: str, limit: int = 100) -> list[NormalizedNewsItem]:
        """Get items by source."""
        return self.query("source = ?", [source], limit)

    def get_alerts(self, limit: int = 100) -> list[NormalizedNewsItem]:
        """Get alert items."""
        conn = self._get_conn()
        cursor = conn.execute("""
            SELECT * FROM news_items
            WHERE json_extract(metadata, '$.is_alert') = 1
            ORDER BY published_at DESC
            LIMIT ?
        """, [limit])
        return [self._row_to_item(row) for row in cursor]

    def close(self) -> None:
        """Close database connection."""
        if self._conn:
            self._conn.close()
            self._conn = None

    def get_path(self) -> str:
        """Get database path."""
        return self.db_path

    def _row_to_item(self, row: sqlite3.Row) -> NormalizedNewsItem:
        """Convert database row to NormalizedNewsItem."""
        metadata_dict = json.loads(row["metadata"] or "{}")

        return NormalizedNewsItem(
            id=row["id"],
            source=row["source"],
            url=row["url"],
            title=row["title"],
            published_at=row["published_at"],
            fetched_at=row["fetched_at"],
            authors=json.loads(row["authors"] or "[]"),
            summary=row["summary"] or "",
            content_text=row["content_text"] or "",
            tickers=json.loads(row["tickers"] or "[]"),
            topics=json.loads(row["topics"] or "[]"),
            language=row["language"] or "en",
            metadata=NewsMetadata(**metadata_dict),
        )
